### ISSUE TITLE
send timeout

### ISSUE THREAD

**Issue by emgerner-msft** [shazow/urllib3#857]
The [urllib3 timeout docs](http://urllib3.readthedocs.io/en/latest/#timeout) show there's read and connect timeouts, but nothing for write. Per the discussion on [this thread in requests](https://github.com/kennethreitz/requests/issues/3099), it would be awesome if there were a way to control write timeout as well as read. Would it be possible to expose a send timeout that worked like read timeout, but for putting data instead of getting data?

@Lukasa for his input as well

**Comment by Lukasa** [comment]
@kevinburke was the king of timeouts though he may have stopped caring. However, this doesn't seem like it'd be too difficult an extension to the current timeout logic, except that we'd want to work out how a send timeout would interact with the `total` timeout flag.

**Comment by emgerner-msft** [comment]
Hey folks, any thoughts on this?

**Comment by shazow** [comment]
Hm I think @kevinburke has moved on from urllib3, so we need someone new to take a look at this. :)

Any chance you're interested in digging into this, @emgerner-msft?

**Comment by emgerner-msft** [comment]
I honestly don't have the time right now. :/ I can poke around a bit more but really learning the urllib3 library and making a good recommendation for a feature at this level is probably not in scope for me.

**Comment by shazow** [comment]
If you find the time, we'd be happy to point you in the right direction. :)

**Comment by emgerner-msft** [comment]
Haha okay, assuming I somehow do, do you want to give me a quick intro of where to look in your code, where I'd probably pipe this feature from (etc)? If at some point I have more time (or someone stumbles on this thread and does) it will be waiting for me.

**Comment by shazow** [comment]
Absolutely.

[connectionpool.py](https://github.com/shazow/urllib3/blob/master/urllib3/connectionpool.py) (and probably connection.py) is where most of the related meat is. I'd search for where various *Timeout exceptions get raised to get a high-level idea.

I don't know off the top of my head, but I *suspect* the challenge might be that httplib (underlying stdlib we depend on) doesn't expose enough granularity to detect the send interval. In that case, we might need to further extend/override things to make this possible.

Feel free to open a WIP pull request and we'll be happy to give more feedback/pointers as you go. :)

**Comment by haikuginger** [comment]
Looking at it, I think we'd need to implement inside `connection.py`. `HTTPConnection.send()` is where the request is getting broken up into individual 8kB segments for transmission.

We could implement a timer there, but it wouldn't happen in realtime - we'd be checking after every 8kB transmission (when we get control back from the socket) if the allotted send time had expired. I would imagine that for most use cases, that'd be "good enough for government work". We wouldn't be able to say "the connection will time out after _n_ seconds", but we would be able to say "the connection will time out the first time we send data to the socket after _n_ seconds have expired".

Of course, if `socket` is doing substantial buffering, then the `send` timeout would become increasingly inaccurate.

Thoughts?  @sigmavirus24, your expertise would be welcome. If this is a reasonable solution, then I can implement it.

**Comment by Lukasa** [comment]
> We could implement a timer there, but it wouldn't happen in realtime - we'd be checking after every 8kB transmission (when we get control back from the socket) if the allotted send time had expired.

@haikuginger That's not accurate: we can set the socket timeout, just as we do for reads.

**Comment by haikuginger** [comment]
I should have known that, but did not - thanks for the correction!

So, if I'm reading this right, setting the socket timeout in `ConnectionPool._make_request()` is all we need to do in the base case (non-chunked request, data being posted has no `.read()` attribute). I believe we'd want to use either the default connect timeout, or the lesser of the explicit `connect` timeout and the explicit `send` timeout if either is set.

For chunked requests, I _believe_ we'll need to maintain a timer that tracks elapsed time and sets the socket timeout lower after each chunk, since the socket would reset the timeout on each operation.

Sound closer?

Side note: It looks like the socket timeout behavior has changed in Python 3.5; whereas previously the timeout reset every time data crossed the socket, it appears that the timeout is now a total time allowed across the whole operation, so we should bear that in mind.

**Comment by Lukasa** [comment]
> I should have known that, but did not - thanks for the correction!

Heh, no problem. There is no "should have known" here: the urllib3 codebase has lots of nooks and crannies, you can't be expected to be an expert in them all yet! I'm certainly not!

> For chunked requests, I believe we'll need to maintain a timer that tracks elapsed time and sets the socket timeout lower after each chunk, since the socket would reset the timeout on each operation.

So right now I don't think we need to do that: we explicitly advertise that the timeouts are not total. From [the docs](https://urllib3.readthedocs.io/en/latest/helpers.html#urllib3.util.timeout.Timeout):

> In addition, the read and total timeouts only measure the time between read operations on the socket connecting the client and the server, not the total amount of time for the request to return a complete response.

I think it's acceptable to keep the same guarantee for send timeouts as well.

**Comment by kevinburke** [comment]
sorry, i'm not really doing python dev any more :( i'm surprised you can
set it! i thought the write was non blocking, somehow.


--
Kevin Burke
925.271.7005 | kev.inburke.com

**Comment by fviard** [comment]
I just came here today after having checked the source code of connectionpool _make_request.
And I would like to ask the same thing as this ticket but from another point of view:
Currently, the "connect" timeout covers:
- self._validate_conn(conn)
- conn.request
So it is not really a "connect" timeout but a "connect+send" timeout.

In my case, I have a situation where I try to upload data (big files) to a https server, but once in a while (20%), the connection is stuck during establishment as it looks like that the server discard the connection without warning.

So, for me, it would be great to have a real "connect" timeout that will only cover tcp connection establishment and eventually ssl negotiation but not the "send/upload" part.
The use case is that any ConnectionTimeout error should be retry-able automatically without issue. At the opposite, once you have started to send data, you can't be sure that you can retry again the same request.
I would like to be able to say "connect" timeout = 3.05s, and "write"(or read timeout)=30s.
But currently I'm needed to have "connect" timeout = 30s, as it covers the real connect and the "upload"/"write" parts.

**Comment by Lukasa** [comment]
@fviard We all agree on the usefulness of the change, we just need someone to do the work. =)

**Comment by tsnoam** [comment]
@Lukasa 
I fixed the code which now passes unitests (besides the GAE build, which I guess from your other builds is an expected failure).
Now, if you have more unitests in mind, I'd be happy to add them.

**Comment by jegoul** [comment]
That change would help solve some problems I had with ssl.

If you do a GET with a small read timeout value,  `connectionpool._make_request` will set the socket timeout to that value before receiving the response.  If you follow this GET with a POST that sends lots of data, you will get an exception with message `'The write operation timed out'`.  That's because ssl uses the socket timeout value (which was set in the previous GET) as its 'time budget' for sending data.

My workaround is to do the first GET with a large read timeout value and then do the POST.

**Comment by ragazzojp** [comment]
Is this problem still applicable after so many years? I'd say yes because I can reproduce cases where uploading large files using _requests_ and therefore _urllib3_ fails with `Time write operation timed out`.

**Comment by sigmavirus24** [comment]
> Is this problem still applicable after so many years?

No one was saying it wasn't applicable. That's why the issue is still open. As the constructive comment here says "We all agree on the usefulness of the change, we just need someone to do the work. =)"

**Comment by ragazzojp** [comment]
I'm sorry, sometimes problems are fixed and issues left pending, especially if very old.

I can try to contribute with a fix, but before coding it I would specify better the overall behavior that is wanted, because IMHO it's this lack of specification that is causing the problem. I would specify explicitly that that `connect_timeout` is only the max time to establish a TCP connection, including TLS handshake. For connections kept open and reuased, this is checked only once, when the connection is established. `read_timeout` is time to receive the first byte of the response after the request is sent. No timeout exception should be raised simply because requests or responses are large and naturally take time to be transmitted.

**Comment by nickwilliams-zaxiom** [comment]
Piping in, because we've encountered this issue, too, and I'd like to further define the behavior that I would expect from this improvement. I might even try to implement this improvement if I can get a 👍  that there's agreement about the expected behavior, but I'm assuming it would have to be exclusive to 2.0.x and couldn't go in the 1.26.x series. I'd love to find out I'm wrong about this...

First, the `Timeout` class. I'd expect it would get an additional `write: _TYPE_TIMEOUT = _DEFAULT_TIMEOUT` argument. It would also get an additional `@property write_timeout` method. This would behave more like `read_timeout` than `connect_timeout`, as it would have to account for the total and the elapsed time like `read_timeout` does.

Next is how the `Timeout` object is used. I expect `connect_timeout` to cover:

* Establishing the connection *or* retrieving the connection from the pool if an existing connection is found, plus
* Validating the connection, plus
* Sending the first 8KB.

Why the third bullet? Because if this is a re-used connection, we may not know if we have a problem with the connection until we try to send the request data. In the quintessential file upload example, we may want a 15 second connect timeout and a 5 minute write timeout for a 1GB file upload, but we don't want it blocking for 5 minutes trying to send the first 8KB of the request. If it can't send 8KB in 15 seconds, we want it to time out early. Hence, the third bullet.

Then, I expect `write_timeout` to cover the sending of everything after the first 8KB (and not apply at all if the request is < 8KB).  An argument could be made that there should be an overlap between `connect_timeout` and `write_timeout`, and that both timeouts should be elapsing while the first 8KB are sent, but in the working case that makes only milliseconds' difference, and in the failing case a sane use case will have a longer `write_timeout` than a `connect_timeout` if separate timeouts are actually needed. So, I don't think overlapping these is actually necessary. Finally, if there is no explicitly configured `write_timeout`, then whatever remains from the `connect_timeout` should be used as the `write_timeout`, which should preserve existing behavior / backwards compatibility (this can probably be covered by the implementation of `@property write_timeout`).

Finally, I expect `read_timeout` to behave exactly as it does now.

What does everyone still paying attention to this issue think? Have I covered everything? Does anyone disagree with any of my expectations, assumptions, or assertions?

Thanks.
